subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/mbox.h>
+
+#include <ipc/ipc_service_backend.h>
+#include "ipc_icmsg.h"
+
+#define DT_DRV_COMPAT	zephyr_ipc_icmsg
+#define CB_BUF_SIZE	CONFIG_IPC_SERVICE_BACKEND_ICMSG_CB_BUF_SIZE
+
+static const uint8_t magic[] = {0x45, 0x6d, 0x31, 0x6c, 0x31, 0x4b,
+				0x30, 0x72, 0x6e, 0x33, 0x6c, 0x69, 0x34};
+BUILD_ASSERT(sizeof(magic) <= CB_BUF_SIZE);
+BUILD_ASSERT(CB_BUF_SIZE <= UINT8_MAX);
+
+struct backend_data_t {
+	/* Tx/Rx buffers. */
+	struct icmsg_buf *tx_ib;
+	struct icmsg_buf *rx_ib;
+
+	/* Backend ops for an endpoint. */
+	const struct ipc_service_cb *ops;
+
+	/* General */
+	struct k_work mbox_work;
+	atomic_t state;
+};
+
+struct backend_config_t {
+	uintptr_t tx_shm_addr;
+	uintptr_t rx_shm_addr;
+	size_t tx_shm_size;
+	size_t rx_shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *dev_data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	static uint8_t cb_buffer[CB_BUF_SIZE];
```

Comment:
Could be. I expect it is me - I don't like placing large buffers on the stack.

@emob-nordic since we are at it. Make this buffer word aligned.

Reply:
No strong opinion here. I think it could be more efficient to put it on stack as long as we know exactly which stack is being used. I would prefer `static` if this function was called from IRQ context, because than we don't have full control which stack is being used.

Reply:
No strong opinion here. I think it could be more efficient to put it on stack as long as we know exactly which stack is being used. I would prefer `static` if this function was called from IRQ context, because than we don't have full control which stack is being used.

Reply:
I will follow your suggestions, remove static and align the buffer.